mitiq/benchmarks/mirror_circuits.py

```diff
@@ -0,0 +1,160 @@
+        num = random.randint(connectivity_graph.size())
+        edges = list(connectivity_graph.edges)
+        curr_edge = edges[num]
+        candidate_edges.add_edge(*curr_edge)
+        connectivity_graph.remove_nodes_from(curr_edge)
+
+    for edge in candidate_edges.edges:
+        if random.random() < two_qubit_gate_prob:
+            final_edges.add_edge(*edge)
+    return final_edges
+
+
+def random_cliffords(connectivity_graph: nx.Graph) -> cirq.Circuit:
+    """Returns a circuit with a two-qubit Clifford gate applied
+    to each edge in edges, and a random single-qubit
+    Clifford gate applied to every other qubit.
+
+    Args:
+        edges: A graph with the edges for which the
+            two-qubit Clifford gate is to be applied.
+    """
+    gates = [
+        cirq.CNOT(cirq.LineQubit(a), cirq.LineQubit(b))
+        for a, b in list(connectivity_graph.edges)
+    ]
+    qubits = nx.Graph()
+    qubits.add_nodes_from(nx.isolates(connectivity_graph))
+    gates.append(random_single_cliffords(qubits))
+    return cirq.Circuit(gates)
+
+
+def random_single_cliffords(connectivity_graph: nx.Graph) -> cirq.Circuit:
+    """Returns a circuit with a random single-qubit Clifford gate
+    applied on each given qubit.
+
+    Args:
+        qubits: A graph with each node representing a qubit for
+            which a random single-qubit Clifford gate is to be applied.
+    """
+    gates: List[cirq.Operation] = []
+    for qubit in connectivity_graph.nodes:
+        num = random.randint(len(cliffords))
+        for clifford_gate in cliffords[num]:
+            gates.append(clifford_gate(cirq.LineQubit(qubit)))
+    return cirq.Circuit(gates)
+
+
+def generate_mirror_circuit(
+    nlayers: int,
+    two_qubit_gate_prob: float,
+    connectivity_graph: nx.Graph,
+    seed: Optional[int] = None,
+) -> cirq.Circuit:
+    """Returns a randomized mirror circuit.
+
+    Args:
+        nlayers: The number of random Clifford layers to be generated.
+            two_qubit_gate_prob: Probability of a two-qubit gate being applied.
+            connectivity_graph: The connectivity graph of the backend
+            on which the mirror circuit will be run. This is used
+            to make sure 2-qubit gates are only run on connected qubits.
+    """
+    if not 0 <= two_qubit_gate_prob <= 1:
+        raise ValueError("two_qubit_gate_prob must be between 0 and 1")
+
+    random.seed(seed)
+
+    nqubits = connectivity_graph.number_of_nodes()
+    single_qubit_cliffords = random_single_cliffords(connectivity_graph)
+
+    forward_circuit = cirq.Circuit()
+
+    quasi_inversion_circuit = cirq.Circuit()
+    quasi_inverse_gates = []
+
+    for _ in range(nlayers):
+        forward_circuit.append(random_paulis(nqubits))
+
+        selected_edges = edge_grab(two_qubit_gate_prob, connectivity_graph)
+        edge_graph = nx.Graph()
+        edge_graph.add_nodes_from(connectivity_graph)
+        edge_graph.add_edges_from(selected_edges.edges)
```

Comment:
`edge_grab` returns a graph - why are you making a new one? Can't you just do `forward_circuit.append(random_cliffords(selected_edges))`?